File: sdks/typescript/src/distribution-channel.ts
import { realpathSync } from "node:fs";

/**
 * Channels a Nimbus binary can be distributed through. When Nimbus runs from a
 * package-manager install, the self-updater steps aside so the package manager
 * owns updates.
 *
 * @moduleStability stable
 *
 * Lives in the SDK so the gateway and CLI share one copy: the `cli` package
 * reaches the gateway over IPC only (no source imports), so a shared pure helper
 * must live in a package both may depend on — the same pattern the manifest
 * signer uses.
 */
export type DistributionChannel = "homebrew" | "scoop" | "winget" | "apt" | "yum" | "msi" | "pkg";

const KNOWN_CHANNELS = new Set<DistributionChannel>([
  "homebrew",
  "scoop",
  "winget",
  "apt",
  "yum",
  "msi",
  "pkg",
] satisfies readonly DistributionChannel[]);

export interface ResolveChannelOptions {
  /** Defaults to `process.env`. */
  env?: Record<string, string | undefined>;
  /** Defaults to `process.execPath`. */
  execPath?: string;
  /**
   * Resolves symlinks so a package manager's real install path (e.g. Homebrew's
   * Cellar) is inspected rather than the `bin` symlink that launched the process.
   * Injectable for tests; defaults to `realpathSync`.
   *
   * May throw: a resolver that fails yields the input path unchanged, and that is handled
   * here rather than expected of the function — so an injected resolver gets the same
   * guarantee the default one does.
   */
  realpath?: (p: string) => string;
}

/**
 * §3.1: a resolver that fails yields the input path unchanged.
 *
 * Wrapped at the point of USE rather than inside the default resolver, because the
 * guarantee has to hold for an *injected* resolver too — and a conformance corpus
 * necessarily injects one. It previously lived inside a `safeRealpath` wrapper around the
 * DEFAULT resolver, so the guarantee held in production and failed for every caller who
 * supplied their own; no test caught that, because none injected a throwing resolver.
 *
 * Failing soft is the right behaviour rather than merely the lenient one: a binary whose
 * path cannot be resolved very often still carries the tell-tale segment, so using the
 * unresolved path strictly increases the number of correct answers.
 */
function resolveSafely(execPath: string, realpath: (p: string) => string): string {
  try {
    return realpath(execPath);
  } catch {
    return execPath;
  }
}

function fromEnv(env: Record<string, string | undefined>): DistributionChannel | null {
  const raw = env["NIMBUS_DISTRIBUTION_CHANNEL"];
  if (raw && KNOWN_CHANNELS.has(raw as DistributionChannel)) {
    return raw as DistributionChannel;
  }
  return null;
}

function fromPath(execPath: string, realpath: (p: string) => string): DistributionChannel | null {
  // Resolve symlinks first: package managers expose the binary via a symlink whose
  // own path may not contain the tell-tale Cellar/apps segment.
  const resolved = resolveSafely(execPath, realpath);
  const p = resolved.replaceAll("\\", "/").toLowerCase();
  // Homebrew: macOS `/opt/homebrew/Cellar/...` or `/usr/local/Cellar/...`,
  // Linuxbrew `/home/linuxbrew/.linuxbrew/...`.
  if (p.includes("/cellar/") || p.includes("/.linuxbrew/")) {
    return "homebrew";
  }
  // Scoop: `~/scoop/apps/<app>/...`.
  if (p.includes("/scoop/apps/")) {
    return "scoop";
  }
  return null;
}

/**
 * Resolve the distribution channel this binary was installed through, or `null`
 * for a plain/direct-download install (where the self-updater stays enabled).
 * An explicit `NIMBUS_DISTRIBUTION_CHANNEL` env marker takes precedence over
 * path heuristics; an unknown marker value is ignored.
 */
export function resolveDistributionChannel(
  opts: ResolveChannelOptions = {},
): DistributionChannel | null {
  const env = opts.env ?? process.env;
  const execPath = opts.execPath ?? process.execPath;
  const realpath = opts.realpath ?? realpathSync;
  return fromEnv(env) ?? fromPath(execPath, realpath);
}

/** Human-facing upgrade hint per channel, used by `nimbus update`. */
export function channelUpgradeHint(channel: DistributionChannel): string {
  switch (channel) {
    case "homebrew":
      return "Installed via Homebrew — run 'brew upgrade nimbus' to update.";
    case "scoop":
      return "Installed via Scoop — run 'scoop update nimbus' to update.";
    case "winget":
      return "Installed via winget — run 'winget upgrade NimbusAgent.Nimbus' to update.";
    case "apt":
      return "Installed via apt — run 'sudo apt update && sudo apt upgrade nimbus' to update.";
    case "yum":
      return "Installed via dnf/yum — run 'sudo dnf upgrade nimbus' to update.";
    case "msi":
      return "Installed via the Windows installer — download the latest .msi from the releases page.";
    case "pkg":
      return "Installed via the macOS installer — download the latest .pkg from the releases page.";
  }
}

File: sdks/go/distributionchannel/distributionchannel.go
package distributionchannel

import (
	"os"
	"path/filepath"
	"strings"
)

// Channel is one of §1's seven distribution channels.
//
// A defined type rather than a bare string, so a caller cannot pass an arbitrary value
// where a channel is expected. The set is closed: a binding MUST NOT accept, resolve to,
// or emit a hint for anything outside it.
type Channel string

// §1's closed set.
const (
	Homebrew Channel = "homebrew"
	Scoop    Channel = "scoop"
	Winget   Channel = "winget"
	Apt      Channel = "apt"
	Yum      Channel = "yum"
	MSI      Channel = "msi"
	Pkg      Channel = "pkg"
)

// EnvVar is §2's marker. Exported because a caller assembling an environment for Resolve
// needs to name it, and hard-coding the string at every call site is how it drifts.
const EnvVar = "NIMBUS_DISTRIBUTION_CHANNEL"

var knownChannels = map[string]Channel{
	"homebrew": Homebrew,
	"scoop":    Scoop,
	"winget":   Winget,
	"apt":      Apt,
	"yum":      Yum,
	"msi":      MSI,
	"pkg":      Pkg,
}

// §4, verbatim. Contract text rather than merely its meaning: a binding returning the right
// advice in different words does not conform. The separator is an em dash (U+2014) and the
// quotes are ASCII apostrophes (U+0027); yum's text names dnf/yum and its command is dnf,
// because the channel is named for the ecosystem while the advice names the tool people use.
var hints = map[Channel]string{
	Homebrew: "Installed via Homebrew — run 'brew upgrade nimbus' to update.",
	Scoop:    "Installed via Scoop — run 'scoop update nimbus' to update.",
	Winget:   "Installed via winget — run 'winget upgrade NimbusAgent.Nimbus' to update.",
	Apt:      "Installed via apt — run 'sudo apt update && sudo apt upgrade nimbus' to update.",
	Yum:      "Installed via dnf/yum — run 'sudo dnf upgrade nimbus' to update.",
	MSI:      "Installed via the Windows installer — download the latest .msi from the releases page.",
	Pkg:      "Installed via the macOS installer — download the latest .pkg from the releases page.",
}

// Config carries the three inputs §R1 requires to be injectable.
//
// The zero value reads the real process: Env from os.Environ, ExecPath from os.Executable,
// and Realpath from filepath.EvalSymlinks. Those defaults are deliberately outside the
// conformance corpus — a case whose expected answer is "whatever this host happens to be"
// would pin nothing.
type Config struct {
	// Env is §2's environment. A nil map is an empty environment, NOT the process's own:
	// a caller who supplies a Config at all has asked to be isolated from the host, and
	// silently reading os.Environ for a nil field would defeat that at the worst moment.
	Env map[string]string

	// ExecPath is the running executable's path. Empty means "read the real one".
	ExecPath string

	// Realpath resolves symlinks. Nil means filepath.EvalSymlinks.
	//
	// It may return an error: §3.1 requires a failure to yield the input path unchanged,
	// and that is handled by this package rather than expected of the function.
	Realpath func(string) (string, error)

	// envSet records whether Env was supplied, so a deliberately empty environment is
	// distinguishable from an absent one. Unexported: a caller sets it by using
	// WithEnv rather than by assignment.
	envSet bool
}

// WithEnv returns a copy of c whose environment is env, including when env is empty.
//
// Needed because Go cannot distinguish "field omitted" from "field set to the zero value",
// and §1 requires an EMPTY environment to be expressible — several corpus cases supply one
// to prove the host is not consulted.
func (c Config) WithEnv(env map[string]string) Config {
	c.Env = env
	c.envSet = true
	return c
}

// resolveSafely applies the resolver, or returns the input unchanged when it fails.
//
// §3.1. Failing soft is right rather than merely lenient: a binary whose path cannot be
// resolved very often still carries the tell-tale segment, so using the unresolved path
// strictly increases the number of correct answers.
func resolveSafely(execPath string, realpath func(string) (string, error)) string {
	resolved, err := realpath(execPath)
	if err != nil {
		return execPath
	}
	return resolved
}

// fromEnv applies §2. An unrecognised value is IGNORED, not an error and not an absence in
// its own right: resolution falls through to §3.
func fromEnv(env map[string]string) (Channel, bool) {
	// Exact string equality: no trimming, no case folding, no aliasing. "apt" matches;
	// "APT", " apt" and "apt-get" do not.
	channel, ok := knownChannels[env[EnvVar]]
	return channel, ok
}

// fromPath applies §3.
func fromPath(execPath string, realpath func(string) (string, error)) (Channel, bool) {
	resolved := resolveSafely(execPath, realpath)
	// strings.ReplaceAll, NOT filepath.ToSlash: ToSlash replaces os.PathSeparator, which on
	// Linux is already "/", so it is a no-op there and a Windows path keeps its
	// backslashes. It converts correctly on Windows, so the mistake passes on a
	// developer's machine and fails in CI's Linux conformance job.
	normalised := strings.ToLower(strings.ReplaceAll(resolved, `\`, "/"))
	if strings.Contains(normalised, "/cellar/") || strings.Contains(normalised, "/.linuxbrew/") {
		return Homebrew, true
	}
	if strings.Contains(normalised, "/scoop/apps/") {
		return Scoop, true
	}
	// §3.2: only these two are path-detectable. A binding MUST NOT add a heuristic for the
	// other five — a new one would make two bindings answer differently for one path.
	return "", false
}

// Resolve reports the channel this binary was installed through (§5).
//
// The environment marker wins outright over the path heuristics. The second return value
// is false for the plain direct-download install, which is a normal answer rather than a
// failure — never an error (§R6).
func Resolve(cfg Config) (Channel, bool) {
	env := cfg.Env
	if !cfg.envSet && env == nil {
		env = environMap()
	}

	execPath := cfg.ExecPath
	if execPath == "" {
		if actual, err := os.Executable(); err == nil {
			execPath = actual
		}
	}

	realpath := cfg.Realpath
	if realpath == nil {
		realpath = filepath.EvalSymlinks
	}

	if channel, ok := fromEnv(env); ok {
		return channel, true
	}
	return fromPath(execPath, realpath)
}

// environMap reads the real process environment, for the zero-value Config only.
func environMap() map[string]string {
	env := map[string]string{}
	for _, entry := range os.Environ() {
		if name, value, found := strings.Cut(entry, "="); found {
			env[name] = value
		}
	}
	return env
}

// UpgradeHint returns §4's advice for a channel, as contract text.
//
// The second return value is false for a value outside §1's closed set. Go checks no
// exhaustiveness on a map lookup and has no non-empty string type, so an unknown Channel is
// reachable however the type is declared — reporting it beats inventing an eighth string.
func UpgradeHint(channel Channel) (string, bool) {
	hint, ok := hints[channel]
	return hint, ok
}
